identity/app/controllers/EmailVerificationController.scala

```diff
@@ -24,28 +25,84 @@ class EmailVerificationController(
     with ImplicitControllerExecutionContext
     with SafeLogging {
 
-  import authenticatedActions.fullAuthWithIdapiUserAction
-
   def completeRegistration(): Action[AnyContent] =
-    fullAuthWithIdapiUserAction.async { implicit request =>
-      val idRequest = idRequestParser(request)
+    Action.async { implicit request =>
       val page = IdentityPage("/complete-registration", "Complete Signup", isFlow = true)
       val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
 
-      Future.successful(
-        Ok(
-          IdentityHtmlPage.html(
-            views.html.verificationEmailResent(
-              request.user,
-              idRequest,
-              idUrlBuilder,
-              verifiedReturnUrlAsOpt,
-              returnUrlVerifier.defaultReturnUrl,
-              isSignupFlow = true,
-            ),
-          )(page, request, context),
-        ),
-      )
+      request
+        .getQueryString("encryptedEmail")
+        .map(encryptedEmail => {
+          api.decryptEmailToken(encryptedEmail) map {
+            case Left(errors) =>
+              logger.error(s"Could not decrypt email address on complete registration page: $errors")
+              errorPage(verifiedReturnUrlAsOpt, page)
+            case Right(email) =>
+              successPage(verifiedReturnUrlAsOpt, page, Some(email))
+                .withSession("encryptedEmail" -> encryptedEmail, "email" -> email)
+          }
+        })
+        .getOrElse({
+          logger.error("No encryptedEmail parameter present on complete registration page request")
+          Future.successful(errorPage(verifiedReturnUrlAsOpt, page))
+        })
+    }
+
+  def resendValidationEmail(): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val page = IdentityPage("/resend-validation", "Resend Validation", isFlow = true)
+      val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
+      val email = request.session.get("email")
+
+      request.session
+        .get("encryptedEmail")
+        .map(token =>
+          api.resendEmailValidationEmailByToken(token, verifiedReturnUrlAsOpt) map {
```

Comment:
just checking, do we have email rate limiting behind this idapi endpoint?

Reply:
yes, we applied the same rate limit as the password reset email - https://github.com/guardian/identity/commit/b37a43411f4e5d4b164f6d3e5d0fd564b03fe9aa